<a href="https://colab.research.google.com/github/ccyang8792/ccyang8792.github.io/blob/main/Linear_Regression_Lab.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##1

In [ ]:
#carsData

In [ ]:
#carsData.info()

In [ ]:
a=carsData.select_dtypes(include='object').columns.tolist()

In [ ]:
b=carsData.select_dtypes(include=['float64','int64']).columns.tolist()

In [ ]:
object_col=carsData[a]

In [ ]:
#object_col

In [ ]:
obj_col=object_col.drop(columns='CarName')

In [ ]:
#obj_col

In [ ]:
cat_col=pd.get_dummies(obj_col, drop_first=True)
cat_col

fueltype_gas  aspiration_turbo  doornumber_two  carbody_hardtop  \
0            True             False            True            False   
1            True             False            True            False   
2            True             False            True            False   
3            True             False           False            False   
4            True             False           False            False   
..            ...               ...             ...              ...   
200          True             False           False            False   
201          True              True           False            False   
202          True             False           False            False   
203         False              True           False            False   
204          True              True           False            False   

     carbody_hatchback  carbody_sedan  carbody_wagon  drivewheel_fwd  \
0                False          False          False           False   
1                False          False          False           False   
2                 True          False          False           False   
3                False           True          False            True   
4                False           True          False           False   
..                 ...            ...            ...             ...   
200              False           True          False           False   
201              False           True          False           False   
202              False           True          False           False   
203              False           True          False           False   
204              False           True          False           False   

     drivewheel_rwd  enginelocation_rear  ...  cylindernumber_three  \
0              True                False  ...                 False   
1              True                False  ...                 False   
2              True                False  ...                 False   
3             False                False  ...                 False   
4             False                False  ...                 False   
..              ...                  ...  ...                   ...   
200            True                False  ...                 False   
201            True                False  ...                 False   
202            True                False  ...                 False   
203            True                False  ...                 False   
204            True                False  ...                 False   

     cylindernumber_twelve  cylindernumber_two  fuelsystem_2bbl  \
0                    False               False            False   
1                    False               False            False   
2                    False               False            False   
3                    False               False            False   
4                    False               False            False   
..                     ...                 ...              ...   
200                  False               False            False   
201                  False               False            False   
202                  False               False            False   
203                  False               False            False   
204                  False               False            False   

     fuelsystem_4bbl  fuelsystem_idi  fuelsystem_mfi  fuelsystem_mpfi  \
0              False           False           False             True   
1              False           False           False             True   
2              False           False           False             True   
3              False           False           False             True   
4              False           False           False             True   
..               ...             ...             ...              ...   
200            False           False           False             True   
201            False           Fa

In [ ]:
#cat_col.info()

In [ ]:
number_col=carsData[b]
number_col=number_col.drop(columns=['car_ID','symboling'])
#number_col

In [ ]:
carsData2=cat_col.join(number_col)

In [ ]:
#carsData2

In [ ]:
#carsData.loc[0]

In [ ]:
#carsData2.loc[0]

In [ ]:
carsData2.head()

fueltype_gas  aspiration_turbo  doornumber_two  carbody_hardtop  \
0          True             False            True            False   
1          True             False            True            False   
2          True             False            True            False   
3          True             False           False            False   
4          True             False           False            False   

   carbody_hatchback  carbody_sedan  carbody_wagon  drivewheel_fwd  \
0              False          False          False           False   
1              False          False          False           False   
2               True          False          False           False   
3              False           True          False            True   
4              False           True          False           False   

   drivewheel_rwd  enginelocation_rear  ...  curbweight  enginesize  \
0            True                False  ...        2548         130   
1            True                False  ...        2548         130   
2            True                False  ...        2823         152   
3           False                False  ...        2337         109   
4           False                False  ...        2824         136   

   boreratio  stroke  compressionratio  horsepower  peakrpm  citympg  \
0       3.47    2.68               9.0         111     5000       21   
1       3.47    2.68               9.0         111     5000       21   
2       2.68    3.47               9.0         154     5000       19   
3       3.19    3.40              10.0         102     5500       24   
4       3.19    3.40               8.0         115     5500       18   

   highwaympg    price  
0          27  13495.0  
1          27  16500.0  
2          26  16500.0  
3          30  13950.0  
4          22  17450.0  

[5 rows x 43 columns]

In [ ]:
#carsData2.info()

In [ ]:
num_cols2=carsData2.select_dtypes(include=['float64','int64']).columns.tolist()
num_cols2.remove('price')
num_cols2

['wheelbase',
 'carlength',
 'carwidth',
 'carheight',
 'curbweight',
 'enginesize',
 'boreratio',
 'stroke',
 'compressionratio',
 'horsepower',
 'peakrpm',
 'citympg',
 'highwaympg']

In [ ]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
carsData2[num_cols2] = scaler.fit_transform(carsData2[num_cols2])
carsData2.head()

fueltype_gas  aspiration_turbo  doornumber_two  carbody_hardtop  \
0          True             False            True            False   
1          True             False            True            False   
2          True             False            True            False   
3          True             False           False            False   
4          True             False           False            False   

   carbody_hatchback  carbody_sedan  carbody_wagon  drivewheel_fwd  \
0              False          False          False           False   
1              False          False          False           False   
2               True          False          False           False   
3              False           True          False            True   
4              False           True          False           False   

   drivewheel_rwd  enginelocation_rear  ...  curbweight  enginesize  \
0            True                False  ...   -0.014566    0.074449   
1            True                False  ...   -0.014566    0.074449   
2            True                False  ...    0.514882    0.604046   
3           False                False  ...   -0.420797   -0.431076   
4           False                False  ...    0.516807    0.218885   

   boreratio    stroke  compressionratio  horsepower   peakrpm   citympg  \
0   0.519071 -1.839377         -0.288349    0.174483 -0.262960 -0.646553   
1   0.519071 -1.839377         -0.288349    0.174483 -0.262960 -0.646553   
2  -2.404880  0.685946         -0.288349    1.264536 -0.262960 -0.953012   
3  -0.517266  0.462183         -0.035973   -0.053668  0.787855 -0.186865   
4  -0.517266  0.462183         -0.540725    0.275883  0.787855 -1.106241   

   highwaympg    price  
0   -0.546059  13495.0  
1   -0.546059  16500.0  
2   -0.691627  16500.0  
3   -0.109354  13950.0  
4   -1.273900  17450.0  

[5 rows x 43 columns]

In [ ]:
#carsData2.info()

In [ ]:
carsTrain2, carsTest2 = train_test_split(carsData2, test_size=0.2, random_state=20)

In [ ]:
model2=LinearRegression()

In [ ]:
carsData2.drop(columns='price')

fueltype_gas  aspiration_turbo  doornumber_two  carbody_hardtop  \
0            True             False            True            False   
1            True             False            True            False   
2            True             False            True            False   
3            True             False           False            False   
4            True             False           False            False   
..            ...               ...             ...              ...   
200          True             False           False            False   
201          True              True           False            False   
202          True             False           False            False   
203         False              True           False            False   
204          True              True           False            False   

     carbody_hatchback  carbody_sedan  carbody_wagon  drivewheel_fwd  \
0                False          False          False           False   
1                False          False          False           False   
2                 True          False          False           False   
3                False           True          False            True   
4                False           True          False           False   
..                 ...            ...            ...             ...   
200              False           True          False           False   
201              False           True          False           False   
202              False           True          False           False   
203              False           True          False           False   
204              False           True          False           False   

     drivewheel_rwd  enginelocation_rear  ...  carheight  curbweight  \
0              True                False  ...  -2.020417   -0.014566   
1              True                False  ...  -2.020417   -0.014566   
2              True                False  ...  -0.543527    0.514882   
3             False                False  ...   0.235942   -0.420797   
4             False                False  ...   0.235942    0.516807   
..              ...                  ...  ...        ...         ...   
200            True                False  ...   0.728239    0.763241   
201            True                False  ...   0.728239    0.949992   
202            True                False  ...   0.728239    0.878757   
203            True                False  ...   0.728239    1.273437   
204            True                False  ...   0.728239    0.975021   

     enginesize  boreratio    stroke  compressionratio  horsepower   peakrpm  \
0      0.074449   0.519071 -1.839377         -0.288349    0.174483 -0.262960   
1      0.074449   0.519071 -1.839377         -0.288349    0.174483 -0.262960   
2      0.604046  -2.404880  0.685946         -0.288349    1.264536 -0.262960   
3     -0.431076  -0.517266  0.462183         -0.035973   -0.053668  0.787855   
4      0.218885  -0.517266  0.462183         -0.540725    0.275883  0.787855   
..          ...        ...       ...               ...         ...       ...   
200    0.339248   1.666445 -0.336970         -0.162161    0.250533  0.577692   
201    0.339248   1.666445 -0.336970         -0.364062    1.416637  0.367529   
202    1.109571   0.926204 -1.232021         -0.338824    0.757535  0.787855   
203    0.435538  -1.183483  0.462183          3.244916    0.047732 -0.683286   
204    0.339248   1.666445 -0.336970         -0.162161    0.250533  0.577692   

      citympg  highwaympg  
0   -0.646553   -0.546059  
1   -0.646553   -0.546059  
2   -0.953012   -0.691627  
3   -0.186865   -0.109354  
4   -1.106241   -1.273900  
..        ...         ...  
200 -0.340094   -0.400490  
201 -0.953012   -0.837195  
202 -1.106241   -1.128332  
203  0.119594   -0.546059  
204 -0.953012   -0.837195  

[205 rows x 42 columns]

In [ ]:
#carsData2

In [ ]:
#carsTest2

In [ ]:
model2.fit(carsTrain2.drop(columns='price'),carsTrain['price'])
model2.score(carsTest2.drop(columns='price'),carsTest2['price'])

0.8113350649510305

In [ ]:
y_pred=model2.predict(carsTest2.drop(columns='price'))
y_pred[-1]

np.float64(9047.484158791027)

In [ ]:
carsTest2['price'].iloc[-1]

np.float64(9495.0)

In [ ]:
abs(y_pred[-1]-carsTest['price'].iloc[-1])

np.float64(447.51584120897314)

The R^2 that I got from included all of the numeric and categorical columns was 0.81.  The gap between the predicted and actual value of the last row in the Test dataset is 447.52.

##2

In [ ]:
#carsData2.head()

In [ ]:
carsData3=carsData2.astype(float)

In [ ]:
all_features=carsData3.drop(columns='price')
all_features.columns

Index(['fueltype_gas', 'aspiration_turbo', 'doornumber_two', 'carbody_hardtop',
       'carbody_hatchback', 'carbody_sedan', 'carbody_wagon', 'drivewheel_fwd',
       'drivewheel_rwd', 'enginelocation_rear', 'enginetype_dohcv',
       'enginetype_l', 'enginetype_ohc', 'enginetype_ohcf', 'enginetype_ohcv',
       'enginetype_rotor', 'cylindernumber_five', 'cylindernumber_four',
       'cylindernumber_six', 'cylindernumber_three', 'cylindernumber_twelve',
       'cylindernumber_two', 'fuelsystem_2bbl', 'fuelsystem_4bbl',
       'fuelsystem_idi', 'fuelsystem_mfi', 'fuelsystem_mpfi',
       'fuelsystem_spdi', 'fuelsystem_spfi', 'wheelbase', 'carlength',
       'carwidth', 'carheight', 'curbweight', 'enginesize', 'boreratio',
       'stroke', 'compressionratio', 'horsepower', 'peakrpm', 'citympg',
       'highwaympg'],
      dtype='object')

In [ ]:
from statsmodels.stats.outliers_influence import variance_inflation_factor
vif_df = pd.DataFrame()
vif_df["feature"] = all_features.columns
vif_df["VIF"] = [variance_inflation_factor(all_features.values, i) for i in range(all_features.shape[1])]

/usr/local/lib/python3.12/dist-packages/statsmodels/stats/outliers_influence.py:197: RuntimeWarning: divide by zero encountered in scalar divide
  vif = 1. / (1. - r_squared_i)


In [ ]:
over_10=vif_df[vif_df['VIF']>10]
over_10

feature         VIF
0          fueltype_gas  452.942240
4     carbody_hatchback   14.243524
5         carbody_sedan   18.651728
7        drivewheel_fwd   11.138069
8        drivewheel_rwd   14.103393
15     enginetype_rotor         inf
16  cylindernumber_five   15.902735
17  cylindernumber_four   68.920791
18   cylindernumber_six   21.031078
21   cylindernumber_two         inf
24       fuelsystem_idi  203.752695
26      fuelsystem_mpfi   10.123039
29            wheelbase   12.033538
30            carlength   15.233213
31             carwidth   11.034931
33           curbweight   33.414050
34           enginesize   46.990068
37     compressionratio  169.648153
38           horsepower   32.767124
40              citympg   37.225051
41           highwaympg   35.535383

In [ ]:
over_5=vif_df[(vif_df['VIF']>5) & (vif_df['VIF']<10)]
over_5


feature       VIF
6     carbody_wagon  9.579882
11     enginetype_l  5.643697
12   enginetype_ohc  7.008684
13  enginetype_ohcf  7.105914
22  fuelsystem_2bbl  7.201008
35        boreratio  7.808087

In [ ]:
over_10['feature'].tolist()

['fueltype_gas',
 'carbody_hatchback',
 'carbody_sedan',
 'drivewheel_fwd',
 'drivewheel_rwd',
 'enginetype_rotor',
 'cylindernumber_five',
 'cylindernumber_four',
 'cylindernumber_six',
 'cylindernumber_two',
 'fuelsystem_idi',
 'fuelsystem_mpfi',
 'wheelbase',
 'carlength',
 'carwidth',
 'curbweight',
 'enginesize',
 'compressionratio',
 'horsepower',
 'citympg',
 'highwaympg']

In [ ]:
over_5['feature'].tolist()

['carbody_wagon',
 'enginetype_l',
 'enginetype_ohc',
 'enginetype_ohcf',
 'fuelsystem_2bbl',
 'boreratio']

The previous 2 cells list the features that have a VIF of over 10 and between 5 and 10.

##3

In [ ]:
carsData2.columns

Index(['fueltype_gas', 'aspiration_turbo', 'doornumber_two', 'carbody_hardtop',
       'carbody_hatchback', 'carbody_sedan', 'carbody_wagon', 'drivewheel_fwd',
       'drivewheel_rwd', 'enginelocation_rear', 'enginetype_dohcv',
       'enginetype_l', 'enginetype_ohc', 'enginetype_ohcf', 'enginetype_ohcv',
       'enginetype_rotor', 'cylindernumber_five', 'cylindernumber_four',
       'cylindernumber_six', 'cylindernumber_three', 'cylindernumber_twelve',
       'cylindernumber_two', 'fuelsystem_2bbl', 'fuelsystem_4bbl',
       'fuelsystem_idi', 'fuelsystem_mfi', 'fuelsystem_mpfi',
       'fuelsystem_spdi', 'fuelsystem_spfi', 'wheelbase', 'carlength',
       'carwidth', 'carheight', 'curbweight', 'enginesize', 'boreratio',
       'stroke', 'compressionratio', 'horsepower', 'peakrpm', 'citympg',
       'highwaympg', 'price'],
      dtype='object')

In [ ]:
features=['curbweight','enginesize','horsepower','doornumber_two','drivewheel_rwd']
features2=['fuelsystem_idi','compressionratio','fueltype_gas','cylindernumber_two','enginetype_rotor']
y=carsData2['price']
import numpy as np
# Fit your model

model = LinearRegression().fit(carsData2[features], y)


# Get R^2
r2 = model.score(carsData2[features], y)


# n = number of rows, p = number of predictors
n = carsData2[features].shape[0]
p = carsData2[features].shape[1]


# Adjusted R^2
adj_r2 = 1 - (1 - r2) * (n - 1) / (n - p - 1)


print("R^2:", r2)
print("Adjusted R^2:", adj_r2)

R^2: 0.8218913050790935
Adjusted R^2: 0.8174162122418848


In [ ]:
features2=['fuelsystem_idi','compressionratio','fueltype_gas','cylindernumber_two','enginetype_rotor']
y=carsData2['price']
import numpy as np
# Fit your model

model = LinearRegression().fit(carsData2[features2], y)


# Get R^2
r2 = model.score(carsData2[features2], y)


# n = number of rows, p = number of predictors
n = carsData2[features2].shape[0]
p = carsData2[features2].shape[1]


# Adjusted R^2
adj_r2 = 1 - (1 - r2) * (n - 1) / (n - p - 1)


print("R^2:", r2)
print("Adjusted R^2:", adj_r2)

R^2: 0.0535371332378165
Adjusted R^2: 0.029756659198565627


Using features=['curbweight','enginesize','horsepower','doornumber_two','drivewheel_rwd'], my R^2 and Adjusted R^2 was .822 and .817.

Using features2=['fuelsystem_idi','compressionratio','fueltype_gas','cylindernumber_two','enginetype_rotor'], my R^2 and Adjusted R^2 was .0535 and .0298.

##4

In [ ]:
#carsTest2

In [ ]:
from sklearn.linear_model import LassoCV
all_feat=carsData2.drop(columns='price')
# 1. Fit LassoCV
lasso = LassoCV(cv=5, random_state=0).fit(all_feat, y)

# 2. Coefficients table
coef_table = pd.DataFrame({
"feature": all_feat.columns,
"coef": lasso.coef_
})
print(coef_table)


# 3. Lists of selected and dropped features
selected = coef_table[coef_table["coef"] != 0]["feature"]
dropped  = coef_table[coef_table["coef"] == 0]["feature"]


print("Selected features:", list(selected))
print("Dropped features:", list(dropped))


# 4. Generate predictions
y_pred2 = lasso.predict(carsTest2.drop(columns='price'))

                  feature         coef
0            fueltype_gas    -0.000000
1        aspiration_turbo     0.000000
2          doornumber_two     0.000000
3         carbody_hardtop     0.000000
4       carbody_hatchback  -358.574671
5           carbody_sedan     0.000000
6           carbody_wagon    -0.000000
7          drivewheel_fwd   -92.828017
8          drivewheel_rwd   887.296298
9     enginelocation_rear     0.000000
10       enginetype_dohcv    -0.000000
11           enginetype_l    -0.000000
12         enginetype_ohc     0.000000
13        enginetype_ohcf     0.000000
14        enginetype_ohcv    -0.000000
15       enginetype_rotor     0.000000
16    cylindernumber_five     0.000000
17    cylindernumber_four -1818.630425
18     cylindernumber_six     0.000000
19   cylindernumber_three     0.000000
20  cylindernumber_twelve    -0.000000
21     cylindernumber_two     0.000000
22        fuelsystem_2bbl    -0.000000
23        fuelsystem_4bbl     0.000000
24         fuelsystem_idi

Dropped features: ['fueltype_gas', 'aspiration_turbo', 'doornumber_two', 'carbody_hardtop', 'carbody_sedan', 'carbody_wagon', 'enginelocation_rear', 'enginetype_dohcv', 'enginetype_l', 'enginetype_ohc', 'enginetype_ohcf', 'enginetype_ohcv', 'enginetype_rotor', 'cylindernumber_five', 'cylindernumber_six', 'cylindernumber_three', 'cylindernumber_twelve', 'cylindernumber_two', 'fuelsystem_2bbl', 'fuelsystem_4bbl', 'fuelsystem_idi', 'fuelsystem_mfi', 'fuelsystem_mpfi', 'fuelsystem_spdi', 'fuelsystem_spfi', 'wheelbase', 'carlength', 'boreratio', 'citympg', 'highwaympg']

In [ ]:
row_100=carsData2['price'].iloc[99]
row_100

np.float64(8949.0)

In [ ]:
y_pred3=lasso.predict(carsData2.drop(columns='price'))

In [ ]:
row_100_pred=y_pred3[99]
row_100_pred

np.float64(10432.717436501043)

In [ ]:
error_row_100=abs(row_100-row_100_pred)
error_row_100

np.float64(1483.7174365010433)